### PR TITLE
[stm] [toplevel] Move delicate state initialization to the STM (BZ#5556)

### PR DESCRIPTION
**Pull request by ejgallego**
We move delicate library/module installation code to the STM so the
API guarantees that the first state snapshot is correct. That was not
the case in the past, which meant that the STM cache was unsound in
batch mode, however we never used its contents due to not backtracking
to the first state.

This provides quite an improvement in the API, however more work is
needed until the codepath is fully polished.

This is a critical step towards handling the STM functionally.

### CODE REVIEW

File: toplevel/coqtop.ml [hunk 1]
```diff
@@ -665,52 +753,62 @@ let init_toplevel arglist =
         exit 1
       end;
       Flags.if_verbose print_header ();
-      inputstate ();
       Mltop.init_known_plugins ();
       engage ();
+
+      (* This state will be shared by all the documents *)
+      Stm.init_core ();
```
Comment (gares) [inline toplevel/coqtop.ml]: I'm uneasy with the name of this function. I think the name should reflect that it snapshots the father of all initial states, as your comment says. What about `Stm.define_initial_doc_state ()`

**Review by gares — CHANGES_REQUESTED** [review]

**Comment by gares** [comment]
Very good!

File: stm/stm.ml [hunk 2]
```diff
@@ -2354,19 +2365,56 @@ end (* }}} *)
...
+  let load_objs libs =
+    let rq_file (dir, from, exp) =
+      let mp = Libnames.(Qualid (Loc.tag @@ qualid_of_string dir)) in
+      let mfrom = Option.map (fun fr -> Libnames.(Qualid (Loc.tag @@ qualid_of_string fr))) from in
+      Flags.silently (Vernacentries.vernac_require mfrom exp) [mp] in
+    List.(iter rq_file (rev libs))
+  in
+
+  (* We must reset the whole state before creating a document! *)
+  State.unfreeze_global_state (Option.get !init_state);
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply stm/stm.ml]: I see your point, I could do the change, however I maybe didn't worry too much here as the next PR actually removes the state handling from the `State` module. WDYT? I guess we will discuss more next week in person.

File: toplevel/coqtop.ml (hunk 1 above)
Comment (ejgallego) [reply toplevel/coqtop.ml]: You have a point, but umm, it maybe will do more than defining the initial state; I could well see it starting the worker manager etc...

File: stm/stm.ml [hunk 3]
```diff
@@ -2354,19 +2365,56 @@ end (* }}} *)
...
+  | VoDoc ln ->
+    let ldir = Flags.verbosely Library.start_library ln in
+    VCS.set_ldir ldir;
+    set_compilation_hints ln
+  | VioDoc ln ->
+    let ldir = Flags.verbosely Library.start_library ln in
+    VCS.set_ldir ldir;
+    set_compilation_hints ln
+  end;
+  load_objs require_libs;
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply stm/stm.ml]: This is delicate [as the title of the PR says], let me have a look at it with calm.

[the remark this answers is not in the document]
Comment (ejgallego) [reply stm/stm.ml]: @gares sorry I don't follow; we _do_ call `start_library` in the 3 cases, if we don't, then we'd get an anomaly. In fact, the previous version worked just by chance as it recorded the initial state before calling `start_librarỳ [c.f. bugzilla]

I am not familiar with CoqIDE sending a file in init, but indeed, we should support the save use case of course [this is intended to be added to the STM as `save_doc` in a further PR]

[the remark this answers is not in the document]
Comment (ejgallego) [reply stm/stm.ml]: If what you mean is why we call `Declaremods` vs `Library.start_library` this is how it was done before, so no change here.

I fully agree that the code in `Library.start_library` should be refactored to some other places, but I've left this to a future PR, I think that what we want to do there will become clearer once we truly have a `stm_doc` record containing all the info of a document.

File: stm/stm.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ejgallego) [reply stm/stm.ml]: After reviewing this, I think the code is right now and should kept that way. Note that `cur_id` will be set correctly by the call to `State.define` just below to define the right initial state. The thing is that we have "two" initial states, the one for the document and the one for the system, so we need to handle them differently as the latter is out of the `State.*` scope and caching.

Regarding the exports, we need to do it; see what we were doing before in Coqtop! This is better I think.

**Comment by ejgallego** [comment]
Thanks for the comments @gares, after careful consideration of your comments I indeed see no change to be made. For the first comment, the `cur_id` will be correctly set by the call to `State.define`, we just replace the hack in coqtop with a less harmful one.

For the second comment, we preserve the previous calling strategy, further cleanup may be needed but IMHO in later PRs.

File: stm/stm.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (gares) [reply stm/stm.ml]: My point about cur_id is that if it gets out of sync with the current system state, then a request to install a state may be optimized away (if the requested state is = to cur_id). 
  https://github.com/coq/coq/blob/master/stm/stm.ml#L763
so the fact that one can unfreeze a state without updating cur_id is a bug in the API.
The synchronization invariant is kept by other functions like
   https://github.com/coq/coq/blob/master/stm/stm.ml#L764

Setting it to dummy (invalidating it)

File: stm/stm.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (gares) [reply stm/stm.ml]: ok

**Comment by gares** [comment]
IMO the invariant in State is still broken, see my new message.

File: stm/stm.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ejgallego) [reply stm/stm.ml]: I see what you mean, but you have a problem anyways in the sense that the "primitive" state has no corresponding `state_id` to be attached to.

[the remark this answers is not in the document]
Comment (ejgallego) [reply stm/stm.ml]: Anyways I find the `cur_id` "invariant" quite fragile, as in the sense that the STM calls some functions that will modify the state without going by `State.define`; this was indeed what happened in the BZ#5556; [this time it was the toplevel but it had no other choice given the API], the cache was bad, so if you tried to go back to sid 1 you got an anomaly.

[the remark this answers is not in the document]
Comment (gares) [reply stm/stm.ml]: At least put a comment that the exported free/unrfeeze are allowing one to break the invariant of cur_id and put a XXX / TODO / FIXME there.

I know I could never grant that *any* piece of code out there was not altering the systems state when not called via State.define, but this way it is even easier to break it. At least put a warning near the new API

[the remark this answers is not in the document]
Comment (ejgallego) [reply stm/stm.ml]: I see your point; the "global" plan I have in mind is to get rid of `State.define`. Maybe we should agree on whether is that what we want to do.

There are two main (and strong IMHO) reasons to replace `State.define` indeed by the traversal function handling the state itself:
- to improve error handling; now, `State.define` takes care of error handling for a function `unit -> unit` so there is not a lot that can be done to handle errors consistently, and there are quite a few of problems due to that (such as error feedbacks not being generated, or just look at what the STM does to most backtraces)
- to return control of state and caching to the traversal! In fact, it is the same case as for errors. `State.define` has no idea of what should be done with the state, as of course, it depends on whatever the traversal function is doing. Hence, the need for so many optional parameters to propagate information that the traversal has and is lost in the abstraction. If you are always sequential, this is fine, but once you branch some state, reasoning becomes complicated.

**Comment by ejgallego** [comment]
Ok, I did change the interface for the root state as suggested. This should be ready then.

**Comment by ejgallego** [comment]
We can shift the discussion about `State.define` to #1095 which is the place where it does belong IMHO.

**Comment by maximedenes** [comment]
@gares is this now ok?

**Review by gares — APPROVED** [review]

**Comment by maximedenes** [comment]
@ejgallego Unlike the PR description says, this doesn't seem to depend on #1088, does it? I'm asking because I would prefer to merge this one first.

**Comment by ejgallego** [comment]
Yes it does [indirectly], if you look at the diff it removes the commands as now the state cannot be handled from the toplevel in the old low-level fashion.

**Comment by ejgallego** [comment]
I have removed the removal of the deprecated and undocumented options `-inputstate / -outputstate`. So once Travis gives the OK this should be good to merge [and the corresponding bug closed]

**Comment by ejgallego** [comment]
Also you were right that this only did a part of what #1088 does, as this only removed the command line versions of `Write State`, not the vernacular ones.

**Comment by ejgallego** [comment]
Stuck on GeoCoq timeout, otherwise this seems good to go in.

File: toplevel/vernac.ml [hunk 4]
```diff
@@ -247,105 +242,3 @@ and load_vernac ~verbosely ~check ~interactive doc sid file =
 let process_expr doc sid loc_ast =
   checknav_deep loc_ast;
   interp_vernac ~interactive:true ~check:true doc sid loc_ast
-
-let warn_file_no_extension =
```
Comment (maximedenes) [inline toplevel/vernac.ml]: Did this warning go away?

Comment (ejgallego) [reply toplevel/vernac.ml]: No, moved to `coqtop`.

Comment (ejgallego) [reply toplevel/vernac.ml]: This patch should be "semantics" preserving except for the bug fixed. So even non-optimal but non-buggy things are preserved indeed.

Comment (maximedenes) [reply toplevel/vernac.ml]: Sorry, I was tricked by the diff being cut. Thanks!

Comment (ejgallego) [reply toplevel/vernac.ml]: I think it is a very good question, in fact, after this `toplevel/vernac` is almost empty [most of the code there is just workarounds over STM limitations in order to imitate more legacy behavior] and should go away at some point with `coqloop` calling the stm directly in a small loop.

Now, `toplevel/coqtop` has inherited some of the "compile" file part, but this code should also gradually be taken over by other components as they assimilate the save/load document functionality.

**Comment by ejgallego** [comment]
Thanks!

**Comment by ejgallego** [comment]
#1103 is also stuck by GeoCoq timing out and is a good candidate to get in now and minimize conflicts I think.

**Comment by herbelin** [comment]
Hi @ejgallego, I'm trying to make the function `go();;` in the interactive tracer working. Apparently `Coqloop.loop` now takes a `doc` as argument. Is it correct to fix the file `base_include` with something of the form:
```coq
let go () = loop (fst (Stm.(new_doc { doc_type = Interactive (Lib.cwd ());  require_libs = [] })));;
```
or is it too crude?
```

**Comment by ejgallego** [comment]
What is the context from where `go` is called? If you mean to preserve the previous state, the above is incorrect. You would need to retrieve the proper previous document and use it.

**Comment by ejgallego** [comment]
An obvious option is to have `loop` return the last used document, then in `console_toploop_run` make it available to the users of `Drop`. Does #6036 help?

**Pull request merged**